docs/docs/config.md
# Quick Server 详细配置

Quick Server 的基本配置都可以由 `config.lua` 完成。除此之外，对于高级用户，还可以直接修改 `nginx.conf` ， `redis.conf` 等配置文件来定制运行环境。

> 在直接编辑 `nginx.conf` 或者 `redis.conf` 时，请自行保证修改后的配置符合 Quick Server 的运行要求，并且与 `config.lua` 中的设置保持兼容。

## config.lua 文件

`config.lua` 文件本身是一个 Lua 的 table。分为应用程序设置和 Quick Server 系统设置两部分。

### 应用程序设置

-   `appRootPath`: 配置用户 app 所在的位置（绝对路径）。例如：`appRootPath = "/home/apps/myapp"`
-   `numOfWorkers`: 指定启动多少个 Nginx Worker 进程。通常，这个数值应该和 CPU 内核数量相等，以获得最佳的性能表现。
-   `appHttpMessageFormat`: 指定 HTTP 请求使用的消息格式，默认为 'json'。目前只支持 'json' 和 'text' 两种。
-   `appSocketMessageFormat`: 指定 Socket 连接使用的消息格式，默认为 'json'。目前只支持 'json' 和 'text' 两种。
-   `appJobMessageFormat`: 指定后台任务存储数据时使用的格式，默认为 'json'。目前只支持 'json' 和 'text' 两种。
-   `appSessionExpiredTime`: 指定 Session 的过期时间，默认为 `600` 秒。


### Quick Server 系统设置

-   `quickserverRootPath`: 指定 Quick Server 的安装目录，默认为 `_QUICK_SERVER_ROOT_`，通常不应修改。
-   `port`: 指定 Quick Server 在哪一个端口接受请求，默认为 `8088`。
-   `welcomeEnabled`: 指定是否启用 Quick Server 自带的欢迎界面，默认为 `true`。在生产服务器上可以将这个选项设置为 `false`，屏蔽掉欢迎界面。
-   `adminEnabled`: 指定是否启用 Quick Server 内置的管理接口，默认为 `true`。欢迎界面中的服务器状态监控等功能需要依赖管理接口。
-   `websocketsTimeout`: WebSocket 协议超时时间，默认值为 `60 * 1000`（60秒）。
-   `websocketsMaxPayloadLen`: 客户端发往服务端的 WebSocket 消息的最大长度，默认值为 `16 * 1024`（16KB）。可以根据自己应用的情况调整这个设置。调整原则是满足需求的前提下，越小越好。
-   `maxSubscribeRetryCount`: 指定订阅广播频道最大尝试次数，默认为 `10` 次。
-   `redis.*`: 指定 Quick Server 内部使用的 Redis 数据库的连接方式和协议超时时间。
-   `beanstalkd.*`: 指定 Quick Server 内部使用的 Beanstalkd 任务队列的连接方式。
-   `beanstalkd.jobTube`: 指定 Quick Server 内部使用什么“桶”来存储后台任务。
-   `monitor.process`: 指定管理接口要监控的进程。
-   `monitor.interval`: 指定监控数据刷新的频率。



CHANGELOG
# Change Log

### 0.6.0

- CHNAGES:
    - `Quick Server` rename to `GameBox Cloud Core`, short name is `gbc-core`.
    - `start_quick_server.sh` rename to `start_server.sh`.
    - `stop_quick_server.sh` rename to `stop_server.sh`.
    - `status_quick_server.sh` rename to `check_server.sh`.
    - the `conf/config.lua` option `quickserverRootPath` rename to `serverRootPath`.

### 0.5.1

- FEATURE:
    - Job Worker is finished as a module of Quick Server.
    - Job Service package is offered. User can use it to add a job easily.

- BUGFIX:
    - start/stop/status shells can exit normally, if some errors happend in progress.
    - start/stop/status shells can get the config of Quick Server in any path.
    - Monitor can show correct number of jobs in Beanstalkd.

- IMPROVE:
    - In error messages, redundant paths can be stripped.
    - Quick Server supports Mac OS now.
    - start/stop/status use more colors for displaying.
    - start/stop/status can show the version and mode of Quick Server.
    - tools.sh can display the result better, even if the result is a lua table.
    - Support job worker module in shells.
    - Add lib luasec-0.5.0.

### 0.5.0

-   UPGRADE:
    -   upgrade luasocket to 3.0-rc1.
    -   add luainspect lib.

-   IMPROVE: make installation better, improve installation shell.
    -   offer an unique install.sh instead of install_ubunutu.sh and install_centos.sh.
    -   except fundamental packages from linux distro, such as "git", "zip" and some building tools, all other packages can be installed without online. This can save much time.
    -   the install.sh supports parameters, ex. user can set the path of installation.

-   IMPROVE: offer all new wiki docs by Sphinx project in ".rst" format.
    -   the new wiki can be outputed in many file formats, such as html, pdf, latex, etc.
    -   it is easy to generated with make tools.
    -   the wiki docs in html format is released with Quick Server, in "docs/" dir.

-   CHNAGE: almost all Quick Server codes are refactored.
    -   new architecture, all modules with "Server" in name are refactored to "Connect".
    -   package mechanism, all functions is offerd as package.
    -   Http and WebSocket modules are refactored, inheriting from "ConnectBase".
    -   new broadcast mechanism.
    -   adjust options in config.lua, more simple, more better.
    -   new maintain and manager shells.
    -   monitor the processes of Quick Server at any time, and the status can be shown in web.

### 0.4.0

-   UPGRADE: From Openresty 1.7.2.x to 1.7.7.x.
-   IMPROVE: make install_ubunutu.sh better in order that user can install Quick Server convenienty, and fix some bugs in it.
    -   create a symbol link for nginx, after Openresty installation.
    -   Add option "--no-check-certificate" for each "wget" command.
    -   "status\_quick\_server.sh" shell file should be copied to installation directory while installation is finished.
    -   Change those shell tools director, move them from  "/opt" to installation path.
    -   Add a new shell tool named "restart\_nginx\_only.sh" in order to restart nginx processes.
    -   Add a parameter for install_ubuntu.sh for specifying installation path insted of absolute path.
    -   Quick Server configure files, including "redis.conf" and "nginx.conf", are modified automatically via "sed" tool.
-   FEATURE: Support plugin mechanism.
    -   Add some methods into "cc" framework, "load" and "bind" etc, for package file(lua file).
    -   Give two plugin examples: "Ranklist" and "ChatRoom", converted from "RanklistAction" and "ChatAction".
    -   Add simple functionality tests for above plugins.
-   IMPROVE: Add a demo action named "HelloworldAction"
    -   include a simple method "/sayhello" to show "hello world".
    -   there are two other methods "addandcount" and "sayletters" to show how to write a function with plugins.
    -   Add "helloworld" client which supports both html and websocket.
-   CHANGE: delete install_mac.sh. The installation of mac env will be supported in next version.
-   OTHER MINOR CHANGES:
    -   IMPROVE: upgrade Quick Server wiki.
    -   BUGFIX: when deploying lua codes defined by user, the target directory in Quick Server shoule be created.
    -   IMPROVE: The target directory in Quick Server can be configured via "luaRepoPrefix" in config.lua for deploying lua codes defined by user.
    -   CHANGE: obsolete the old interface of uploading user codes.
    -   IMPROVE: Add two sql files: "base.sql" and "pre_condition.sql" in conf/sql for configuring MySql.
    -   CHANGE: don't need to set "root" privilege in nginx conf file.
    -   IMPROVE: Add README file for each sub-dir.
    -   CHANGE: change some shells which are used to encapsulate "nginx" command in "openresty/nginx".

### 0.4.0-rc0

-   FEATURE: Add an new action "SessionAction" to generate **session_id** for user.
-   FEATURE: adjust many interfaces in RanklistAction.
    -   Each interface calling needs checking **session_id**.
    -   "Add" interface can generate a uid according to the **nickname** when user calls it first time.
    -   the format of uid is "nickname+numbers" in order to keep each uid unique.
    -   score, remove, getrank and getrevrank should get key from param "uid".
    -   "GetRevRank" also replies score.
    -   "GetRankAction" also return socre.
    -   "AddAction" can return a percent to indicate "rank/total", in other words, it's the user's positiono in a ranklist.
    -   Add some test cases for RanklistAction.
-   BUGFIX: if a redis command fails, it replies an int 0 not a string "0".

### 0.3.9

-   FEATURE: implement a ranklist with friendship based on social network platform.(only support Weibo now)
-   FEATURE: Each interface calling needs checking **session_id**.
-   FEATURE: ADD some test cases for FriendshipAction.
-   BUGFIX: table.length() method in function.lua moduel.

### 0.3.8

-   FEATURE: implement chatting room, ChatAction module.

conf/config.lua
--[[

Copyright (c) 2015 gameboxcloud.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

]]

_GAMEBOX_CLOUD_CORE_VERSION = "0.5.1"

_DBG_ERROR = 0
_DBG_WARN  = 1
_DBG_INFO  = 2
_DBG_DEBUG = 3

local config = {
    -- user app
    appRootPath = "_GBC_CORE_ROOT_/<USER_APP_ROOT>",

    numOfWorkers = 4,

    appHttpMessageFormat   = "json",
    appSocketMessageFormat = "json",
    appJobMessageFormat    = "json",
    appSessionExpiredTime  = 60 * 10, -- 10m

    -- GameBox Cloud Core settings
    serverRootPath = "_GBC_CORE_ROOT_",
    port = 8088,
    welcomeEnabled = true,
    adminEnabled = true,
    websocketsTimeout = 60 * 1000, -- 60s
    websocketsMaxPayloadLen = 16 * 1024, -- 16KB
    maxSubscribeRetryCount = 10,

    -- internal memory database
    redis = {
        socket     = "unix:_GBC_CORE_ROOT_/tmp/redis.sock",
        -- host       = "127.0.0.1",
        -- port       = 6379,
        timeout    = 10 * 1000, -- 10 seconds
    },

    -- background job server
    beanstalkd = {
        host       = "127.0.0.1",
        port       = 11300,
        jobTube    = "jobTube",
    },

    -- internal monitor
    monitor = {
        process = {
            "nginx",
            "redis-server",
            "beanstalkd",
        },

        interval = 2,
    },
}

return config
